### PR TITLE
feat(aws-fargate-sqs): Created new construct

### LINKED ISSUE #587: feat(aws-fargate-sqs): New Construct
Create a construct that links a Fargate Service to an SQS Queue

Should create a VPC or accept an existing VPC
Should create a Fargate Service or accept an existing service
Should create an SQS queue or accept an existing queue
Should grant the service IAM permission to send and delete messages on the queue
Should set up environment variables in the service that identify the SQS queue in the construct code so the container can publish to the correct queue

### Other

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by mickychetta**
*Issue #587 , if available:*

*Description of changes:*
-Created new construct with Fargate connected to SQS
-Added environment variables
-Able to create VPC, service, queue, and accept existing ones

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

fixes #587

### CODE REVIEW

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: f30b9ec81f51f7800075f8f19d30b192c6c7a760
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=79a6775e-5557-497b-b834-572149137458%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: f30b9ec81f51f7800075f8f19d30b192c6c7a760
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2de8ef53-88bc-4fdd-a4ed-4a30ca1f9db4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 7bd27ddafbc0053609947a7d781add05657ea741
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a3382c4b-5830-4b73-82c8-b25416e24443%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 7bd27ddafbc0053609947a7d781add05657ea741
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=eb009b82-5414-4485-9ed2-35222c0ee91c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 645475b67e1cc225ea359c30da7f91338bab1dba
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5cc681d8-a8ed-4da2-af34-4ec678ae55c7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 645475b67e1cc225ea359c30da7f91338bab1dba
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f60a0943-263d-42eb-93bf-2869bb1abab5%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: d64658a4f2250b40a4487bd04c5d7d3196dc6223
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=acc67ac5-8f37-4fa8-b067-70e3f3b1a454%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: d64658a4f2250b40a4487bd04c5d7d3196dc6223
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4513f543-4e54-45e9-98a9-ef71c7dab6d1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 6f1c67ff7df5ebeafc635d1c978e0d85fc7d2779
* Result: FAILED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7977a117-f197-4843-987b-49e39d3971f0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 6f1c67ff7df5ebeafc635d1c978e0d85fc7d2779
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7e7020ab-515b-410c-aea0-2d17a5ca5292%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: f9d5e67814104deb91a400e8720dcf0af9a68e32
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8dacf074-4264-4d9a-879a-bfdd4dc40e99%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: f9d5e67814104deb91a400e8720dcf0af9a68e32
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=54e19f6b-ef8e-44d9-a512-c3f427fb36c7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+  expect(stack).toCountResources('AWS::SQS::Queue', 1);
+  expect(stack).toCountResources('AWS::ECS::Service', 1);
+});
+
+test('Existing service/new queue, public API, existing VPC', () => {
+  // An environment with region is required to enable logging on an ALB
+  const stack = new cdk.Stack(undefined, undefined, {
+    env: { account: "123456789012", region: 'us-east-1' },
+  });
+  const publicApi = true;
```
[the remark this answers is not in the document]
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts:195]: Yes publicApi is a required construct prop which determine the type of subnets to create. 

publicApi = `true`
'''
// Confirm we created a Public/Private VPC
  expect(stack).toHaveResourceLike('AWS::EC2::InternetGateway', {});
  expect(stack).toCountResources('AWS::EC2::VPC', 1);
  expect(stack).toCountResources('AWS::SQS::Queue', 2);
  expect(stack).toCountResources('AWS::ECS::Service', 1);
'''

publicApi = `false`
'''
// Confirm we created an Isolated VPC
  expect(stack).not.toHaveResourceLike('AWS::EC2::InternetGateway', {});
  expect(stack).toCountResources('AWS::EC2::VPC', 1);
  expect(stack).toCountResources('AWS::SQS::Queue', 1);
  expect(stack).toCountResources('AWS::ECS::Service', 1);
'''

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: f191ad2655478fa9ef905d5d10c4caf17136c04f
* Result: FAILED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b9b4d5d7-bd42-4382-8db8-77d72060c190%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: f191ad2655478fa9ef905d5d10c4caf17136c04f
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b17db8b5-661c-4384-9763-5935ff75e6ad%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: f191ad2655478fa9ef905d5d10c4caf17136c04f
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2ce7b14c-e6a6-404f-ba7d-92108a76dbca%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+  const serviceName = "custom-service-name";
+  const queueName = "custom-queue-name";
+
+  new FargateToSqs(stack, 'test-construct', {
+    publicApi,
+    ecrRepositoryArn: defaults.fakeEcrRepoArn,
+    vpcProps: { cidr: '172.0.0.0/16' },
+    clusterProps: { clusterName },
+    containerDefinitionProps: { containerName },
+    fargateTaskDefinitionProps: { family: 'family-name' },
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts:39]: oops - 'family-name' ought to be a constant like those defined in lines 27-31. This is an error in my code in aws-fargate-sqs.

File: source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+  expect(stack).toHaveResourceLike('AWS::EC2::InternetGateway', {});
+  expect(stack).toCountResources('AWS::EC2::VPC', 1);
+  expect(stack).toCountResources('AWS::SQS::Queue', 2);
+  expect(stack).toCountResources('AWS::ECS::Service', 1);
+});
+
+test('New service/new queue, private API, new VPC', () => {
+
+  // An environment with region is required to enable logging on an ALB
+  const stack = new cdk.Stack(undefined, undefined, {
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts:110]: Same comment

File: source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+    ]
+  });
+
+  expect(stack).toHaveResourceLike("AWS::EC2::VPC", {
+    CidrBlock: '172.0.0.0/16'
+  });
+  // Confirm we created a Public/Private VPC
+  expect(stack).toHaveResourceLike('AWS::EC2::InternetGateway', {});
+  expect(stack).toCountResources('AWS::EC2::VPC', 1);
+  expect(stack).toCountResources('AWS::SQS::Queue', 2);
```
Comment (biffgaut) [inline source/patterns/@aws-solutions-constructs/aws-fargate-sqs/test/fargate-sqs.test.ts:103]: We should have one test that specifies DeadLetterQueue props and checks that a queue with the proper name exists. Looking at all our other SQS constructs, it looks like this is something we have missed completely up to this point.

**Review by biffgaut — CHANGES_REQUESTED** [review]
Some questions, one additional test requested (or update an existing test to check the variation), maybe some trivial changes.

File: source/patterns/@aws-solutions-constructs/aws-fargate-sqs/lib/index.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    // Setup the SQS Queue
+    [this.sqsQueue] = defaults.buildQueue(this, `${id}-queue`, {
+      queueProps: props.queueProps,
+      deadLetterQueue: this.deadLetterQueue,
+      existingQueueObj: props.existingQueueObj,
+    });
+
+    // Enable message send and receive permissions for Fargate service by default
+    this.sqsQueue.grantSendMessages(this.service.taskDefinition.taskRole);
+    this.sqsQueue.grantConsumeMessages(this.service.taskDefinition.taskRole);
```
[the remark this answers is not in the document]
Comment (mickychetta) [reply source/patterns/@aws-solutions-constructs/aws-fargate-sqs/lib/index.ts:204]: Did not copy grantConsumeMessages from an existing construct but thinking of different use cases, we should allow users to read and write to the queue

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 74e74c60526e17763edff9d9f5039d85920baff3
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=add526aa-9f2b-4664-b076-6d632848cf5b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: 74e74c60526e17763edff9d9f5039d85920baff3
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=cb2f8e00-ba5f-4c25-8812-1374001c518c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by biffgaut — CHANGES_REQUESTED** [review]
Let's add a prop:

```
|queuePermissions?|`string[]`|Optional queue permissions to grant to the Fargate service. One or more of the following may be specified: `Read`,`Write`. Default is 'Write' |
```

Chose Write for default as a service will be one or the other in most cases so I had to choose one. Since the preferred way to read a queue is a event driven Lambda function, I chose write - but I'm open to discussion.

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: 65a451252ad4b38b4f137be2be6025ce01f8a69a
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3b5a7569-bb1d-4ff3-8f19-dd84f9930f5e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Review by biffgaut — APPROVED** [review]

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v2
* Commit ID: b753cd96143923b8c149fa3b9f791c3a689a4cb6
* Result: SUCCEEDED
* [Build Logs](https://j8dgcooz5f.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e2910298-3b34-4bdf-b4a7-0d98a54119e4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by aws-solutions-constructs-team** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: githubautobuild-for-cdk-v1
* Commit ID: b753cd96143923b8c149fa3b9f791c3a689a4cb6
* Result: SUCCEEDED
* [Build Logs](https://nis3u8ftb6.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f48b2e62-a57a-4a0e-8df6-641ac8a12c7c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**